File: client/cpp/protocols.cpp
#include "secure_message_protocol.h"
#include "crypto.h"
#include <openssl/sha.h> // for SHA256
#include <sstream>
#include <iomanip>
#include <cmath>

namespace anon {

// Static members initialization
std::unordered_set<std::string> SecureMessageProtocol::processed_msg_ids_;
std::mutex SecureMessageProtocol::processed_ids_mutex_;

SecureMessageProtocol::SecureMessageProtocol()
: encryption_engine_(), traffic_manager_() {}

std::string SecureMessageProtocol::generate_message_id(const std::string& plaintext) {
    double now = epoch_seconds_now_double();
    bytes entropy = anon_crypto::secure_random_bytes(8);

    // Build SHA256(plaintext || entropy || timestamp)
    SHA256_CTX ctx;
    SHA256_Init(&ctx);
    SHA256_Update(&ctx, plaintext.data(), plaintext.size());
    SHA256_Update(&ctx, entropy.data(), entropy.size());

    std::ostringstream ts_ss;
    ts_ss << std::fixed << std::setprecision(6) << now;
    std::string ts_str = ts_ss.str();
    SHA256_Update(&ctx, ts_str.data(), ts_str.size());

    unsigned char hash[SHA256_DIGEST_LENGTH];
    SHA256_Final(hash, &ctx);

    // Return first 16 hex chars
    std::ostringstream oss;
    oss << std::hex << std::setfill('0');
    for (int i = 0; i < 8; ++i) { // 16 hex chars = 8 bytes
        oss << std::setw(2) << static_cast<int>(hash[i]);
    }
    return oss.str();
}

std::optional<bytes> SecureMessageProtocol::encrypt_message(
    const std::string& plaintext,
    const bytes& key,
    const bytes& associated_data
) {
    if (plaintext.empty() || key.size() != 32) {
        std::cerr << "[-] Encryption error: Invalid input (empty message or wrong key length).\n";
        return std::nullopt;
    }

    try {
        std::string message_id = generate_message_id(plaintext);

        // Build envelope
        json message_envelope = {
            {"content", plaintext},
            {"timestamp", epoch_seconds_now_double()},
            {"message_id", message_id},
            {"version", "secure_v1"}
        };

        std::string serialized = message_envelope.dump();

        // AES-GCM encryption
        bytes nonce = anon_crypto::secure_random_bytes(12);
        auto ciphertext = anon_crypto::aes_gcm_encrypt(
            key, nonce, bytes(serialized.begin(), serialized.end()), associated_data
        );

        // Return concatenated nonce + ciphertext
        bytes result;
        result.reserve(nonce.size() + ciphertext.size());
        result.insert(result.end(), nonce.begin(), nonce.end());
        result.insert(result.end(), ciphertext.begin(), ciphertext.end());
        return result;

    } catch (const std::exception& e) {
        std::cerr << "[-] Unexpected encryption failure: " << e.what() << "\n";
        return std::nullopt;
    } catch (...) {
        std::cerr << "[-] Unknown encryption failure\n";
        return std::nullopt;
    }
}

std::optional<std::string> SecureMessageProtocol::decrypt_message(
    const bytes& encrypted_data,
    const bytes& key,
    const bytes& associated_data
) {
    if (encrypted_data.size() < 28 || key.size() != 32) {
        std::cerr << "[-] Decryption error: Invalid input (too short or wrong key length).\n";
        return std::nullopt;
    }

    try {
        bytes nonce(encrypted_data.begin(), encrypted_data.begin() + 12);
        bytes ciphertext(encrypted_data.begin() + 12, encrypted_data.end());

        bytes decrypted = anon_crypto::aes_gcm_decrypt(key, nonce, ciphertext, associated_data);

        json message_envelope = json::parse(decrypted);

        if (!message_envelope.contains("content") ||
            !message_envelope.contains("timestamp") ||
            !message_envelope.contains("message_id")) {
            std::cerr << "[-] Decryption error: Missing required fields in envelope.\n";
            return std::nullopt;
        }

        std::string msg_id = message_envelope["message_id"].get<std::string>();

        // Replay attack check
        {
            std::lock_guard<std::mutex> lock(processed_ids_mutex_);
            if (processed_msg_ids_.size() > max_processed_ids) {
                // Keep only recent IDs
                std::unordered_set<std::string> recent;
                size_t count = 0;
                for (const auto& id : processed_msg_ids_) {
                    if (count++ >= recent_keep_count) break;
                    recent.insert(id);
                }
                processed_msg_ids_ = std::move(recent);
            }

            if (processed_msg_ids_.count(msg_id)) {
                std::cerr << "[-] Decryption warning: Possible replay attack (duplicate ID).\n";
                return std::nullopt;
            }
            processed_msg_ids_.insert(msg_id);
        }

        double timestamp = message_envelope["timestamp"].get<double>();
        if (std::abs(epoch_seconds_now_double() - timestamp) > 600.0) {
            std::cerr << "[-] Decryption error: Message timestamp invalid (clock skew or expired).\n";
            return std::nullopt;
        }

        return message_envelope["content"].get<std::string>();

    } catch (const json::exception& e) {
        std::cerr << "[-] Decryption error: Invalid JSON in envelope: " << e.what() << "\n";
        return std::nullopt;
    } catch (const std::exception& e) {
        std::cerr << "[-] Decryption error: " << e.what() << "\n";
        return std::nullopt;
    } catch (...) {
        std::cerr << "[-] Unknown decryption failure\n";
        return std::nullopt;
    }
}

} // namespace anon

File: client/protocols.py
from .encryption import *
from .traffic_manager import *

class SecureMessageProtocol:
    processed_msg_ids = set()

    def __init__(self):
        self.encryption_engine = E2EEncryptionEngine()
        self.traffic_manager = AnonymousTrafficManager()

    @staticmethod
    def encrypt_message(
        plaintext: str, key: bytes, associated_data: bytes = b""
    ) -> Optional[bytes]:
        if not plaintext or not key or len(key) != 32:
            print(
                "[-] Encryption error: Invalid input (empty message or wrong key length)."
            )
            return None
        try:
            current_time = int(time.time())
            entropy = os.urandom(8)
            message_id = hashlib.sha256(
                plaintext.encode() + entropy + str(current_time).encode()
            ).hexdigest()[:16]

            message_envelope = {
                "content": plaintext,
                "timestamp": time.time(),
                "message_id": message_id,
                "version": "secure_v1",
            }

            serialized_message = json.dumps(message_envelope, separators=(",", ":"))
            aesgcm = AESGCM(key)
            nonce = os.urandom(12)
            ciphertext = aesgcm.encrypt(
                nonce, serialized_message.encode(), associated_data
            )

            return nonce + ciphertext

        except ValueError as ve:
            print(f"[-] Encryption error (invalid key/nonce): {ve}")
            return None
        except Exception as enc_err:
            print(f"[-] Unexpected encryption failure: {enc_err}")
            return None

    @staticmethod
    def decrypt_message(
        encrypted_data: bytes, key: bytes, associated_data: bytes = b""
    ) -> Optional[str]:
        if not encrypted_data or len(encrypted_data) < 28 or not key or len(key) != 32:
            print(
                "[-] Decryption error: Invalid input (too short or wrong key length)."
            )
            return None
        try:
            nonce = encrypted_data[:12]
            ciphertext = encrypted_data[12:]

            aesgcm = AESGCM(key)
            decrypted_data = aesgcm.decrypt(nonce, ciphertext, associated_data)
            message_envelope = json.loads(decrypted_data.decode())

            required_fields = ["content", "timestamp", "message_id"]
            if not all(field in message_envelope for field in required_fields):
                print("[-] Decryption error: Missing required fields in envelope.")
                return None

            msg_id = message_envelope["message_id"]

            if len(SecureMessageProtocol.processed_msg_ids) > 1000:
                recent_ids = list(SecureMessageProtocol.processed_msg_ids)[-500:]
                SecureMessageProtocol.processed_msg_ids.clear()
                SecureMessageProtocol.processed_msg_ids.update(recent_ids)

            if msg_id in SecureMessageProtocol.processed_msg_ids:
                print("[-] Decryption warning: Possible replay attack (duplicate ID).")
                return None
            SecureMessageProtocol.processed_msg_ids.add(msg_id)

            current_time = time.time()
            if abs(current_time - message_envelope["timestamp"]) > 600:
                print(
                    "[-] Decryption error: Message timestamp invalid (clock skew or expired)."
                )
                return None

            return message_envelope["content"]

        except json.JSONDecodeError as json_err:
            print(f"[-] Decryption error: Invalid JSON in envelope: {json_err}")
            return None
        except ValueError as dec_ve:
            print(f"[-] Decryption error (auth failure/invalid nonce): {dec_ve}")
            return None
        except Exception as dec_err:
            print(f"[-] Unexpected decryption failure: {dec_err}")
            return None
